### PR TITLE
Tweaks for async / other race fixes from `blend/jobkit`

### PR DESCRIPTION
**Pull request by wcharczuk**
This PR represents a number of subtle changes that can have adverse affects depending on how repositories use some of the `async` and `cron` packages. It is submitted as an advisory PR to start.

### CODE REVIEW

File: async/batch.go [hunk 1]
```diff
@@ -46,27 +46,34 @@ type Batch struct {
...
 		go worker.Start()
-		<-worker.NotifyStarted()
-		workers <- worker
+		<-workerStarted
+
+		allWorkers[x] = worker
+		availableWorkers <- worker
 	}
 
 	defer func() {
```
Comment (wcharczuk) [inline async/batch.go]: note; this caused a pretty heinous `why is this thing frozen` bug in another package

File: async/latch.go [hunk 2]
```diff
@@ -37,6 +37,17 @@ type Latch struct {
 	stopped  chan struct{}
 }
 
+// Reset resets the latch.
+func (l *Latch) Reset() {
```
Comment (wcharczuk) [inline async/latch.go]: important usage note here is you _MUST_ call `.Reset()` if you want to have the latch start after an explicit stop

File: cron/job_builder.go [hunk 3]
```diff
@@ -12,19 +12,21 @@ var (
...
 	_ ShouldSkipLoggerOutputProvider    = (*JobBuilder)(nil)
-	_ OnStartReceiver                   = (*JobBuilder)(nil)
-	_ OnCancellationReceiver            = (*JobBuilder)(nil)
-	_ OnCompleteReceiver                = (*JobBuilder)(nil)
-	_ OnFailureReceiver                 = (*JobBuilder)(nil)
-	_ OnBrokenReceiver                  = (*JobBuilder)(nil)
-	_ OnFixedReceiver                   = (*JobBuilder)(nil)
-	_ OnEnabledReceiver                 = (*JobBuilder)(nil)
-	_ OnDisabledReceiver                = (*JobBuilder)(nil)
+	_ OnBeginHandler                    = (*JobBuilder)(nil)
```
Comment (wcharczuk) [inline cron/job_builder.go]: this is kind of an unnecessary change but shouldn't affect too much

File: cron/job_builder.go [hunk 4]
```diff
@@ -269,45 +309,45 @@ func (jb *JobBuilder) HistoryMaxAge() time.Duration {
 	return jb.Config.HistoryMaxAgeOrDefault()
 }
 
-// OnStart is a lifecycle hook.
-func (jb *JobBuilder) OnStart(ctx context.Context) {
-	if jb.OnStartHandler != nil {
-		jb.OnStartHandler(GetJobInvocation(ctx))
+// OnBegin is a lifecycle hook.
```
Comment (wcharczuk) [inline cron/job_builder.go]: Note/Context: this was changed because the scheduler `Start/Stop`s and jobs `Begin/Complete`

File: cron/job_interfaces.go [hunk 5]
```diff
@@ -37,6 +37,16 @@ type ScheduleProvider interface {
 	Schedule() Schedule
 }
 
+// OnLoadHandler is a job that can run an on load step.
```
Comment (wcharczuk) [inline cron/job_interfaces.go]: The purpose of these is to let you start and stop stateful processes in jobs when the job is loaded or unloaded, this is used by jobkit for notification retry workers

File: cron/job_manager.go [hunk 6]
```diff
@@ -57,16 +56,37 @@ func (jm *JobManager) LoadJobs(jobs ...Job) error {
 			OptJobSchedulerTracer(jm.Tracer),
 			OptJobSchedulerLog(jm.Log),
 		)
-
 		if err := scheduler.RestoreHistory(context.Background()); err != nil {
 			logger.MaybeError(jm.Log, err)
 			continue
 		}
+		if typed, ok := job.(OnLoadHandler); ok {
```
Comment (wcharczuk) [inline cron/job_manager.go]: the diff here is confusing; this addition lands in `LoadJobs(jobs...Job) error`

File: web/static_file_server.go [hunk 7]
```diff
@@ -135,27 +138,35 @@ func (sc *StaticFileServer) ServeCachedFile(r *Ctx, filepath string) Result {
 	if file.ETag != "" {
 		r.Response.Header().Set(webutil.HeaderETag, file.ETag)
 	}
+
+	r.WithContext(logger.WithLabels(r.Context(), logger.Labels{"web.static_file_cached": file.Path}))
```
Comment (wcharczuk) [inline web/static_file_server.go]: we may want to add a more reasonable way to tack on event labels for logging than this, but for now it works

File: cron/job_builder.go (hunk 4 above)
Comment (dhermes) [reply cron/job_builder.go]: (Just deleted a question up above asking why)

File: cron/job_interfaces.go [hunk 8]
```diff
@@ -62,45 +72,45 @@ type DisabledProvider interface {
 	Disabled() bool
 }
 
-// OnStartReceiver is an interface that allows a task to be signaled when it has started.
-type OnStartReceiver interface {
-	OnStart(context.Context)
+// OnBeginHandler is an interface that allows a job invocation to be signaled when it has started.
+type OnBeginHandler interface {
```
Comment (dhermes) [inline cron/job_interfaces.go]: I am partial to `Receiver` over `Handler`, just my two cents

File: cron/job_manager.go [hunk 9]
```diff
@@ -57,16 +56,37 @@ func (jm *JobManager) LoadJobs(jobs ...Job) error {
...
+	for _, jobName := range jobNames {
+		if jobScheduler, ok := jm.Jobs[jobName]; ok {
+			if typed, ok := jobScheduler.Job.(OnUnloadHandler); ok {
+				typed.OnUnload()
+			}
+			jobScheduler.Stop()
+			delete(jm.Jobs, jobName)
+		} else {
+			return ex.New(ErrJobNotFound, ex.OptMessagef("job: %s", jobName))
+		}
```
Comment (dhermes) [inline cron/job_manager.go]: Nit: I try to avoid `else` blocks in these situations for mat ryer's [Align the happy path to the left edge][1] (I'm aware you're aware of this):

```go
jobScheduler, ok := jm.Jobs[jobName]
if !ok {
	return ex.New(ErrJobNotFound, ex.OptMessagef("job: %s", jobName))
}

if typed, ok := jobScheduler.Job.(OnUnloadHandler); ok {
	typed.OnUnload()
}
jobScheduler.Stop()
delete(jm.Jobs, jobName)
```

Do as you wish though

[1]: https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88

File: cron/job_manager.go [hunk 10]
```diff
@@ -226,11 +243,10 @@ func (jm *JobManager) Status() JobManagerStatus {
 
 // Start starts the job manager and blocks.
 func (jm *JobManager) Start() error {
-	stopped := jm.Latch.NotifyStopped()
 	if err := jm.StartAsync(); err != nil {
 		return err
 	}
-	<-stopped
+	<-jm.Latch.NotifyStopped()
```
Comment (dhermes) [inline cron/job_manager.go]: This change is OK because the `l.stopped` channel can't be replaced by `StartAsync()`?

File: cron/job_manager.go [hunk 11]
```diff
@@ -243,8 +259,6 @@ func (jm *JobManager) StartAsync() error {
 	jm.Latch.Starting()
 	logger.MaybeInfo(jm.Log, "job manager starting")
 	for _, job := range jm.Jobs {
-		job.Log = jm.Log
-		job.Tracer = jm.Tracer
```
Comment (dhermes) [inline cron/job_manager.go]: Why remove these?

File: cron/job_manager.go [hunk 12]
```diff
@@ -254,59 +268,25 @@ func (jm *JobManager) StartAsync() error {
...
-		job.Stop()
-	}
-
-	logger.MaybeInfo(jm.Log, "job manager paused")
-	return nil
-}
-
-// Resume restarts the job manager's job schedulers.
-// This call is asynchronous and does not block.
-func (jm *JobManager) Resume() error {
```
Comment (dhermes) [inline cron/job_manager.go]: These features (`Pause()` and `Resume()`) are just being removed?

File: cron/job_manager_test.go [hunk 13]
```diff
@@ -84,24 +84,27 @@ func TestJobManagerJobPanicHandling(t *testing.T) {
 }
 
 func TestEnabledProvider(t *testing.T) {
-	a := assert.New(t)
-	a.StartTimeout(2000 * time.Millisecond)
-	defer a.EndTimeout()
+	assert := assert.New(t)
```
Comment (dhermes) [inline cron/job_manager_test.go]: Variable / import shadowing, a discussion for another day :)

File: cron/util.go [hunk 14]
```diff
@@ -95,3 +95,31 @@ func IsWeekDay(day time.Weekday) bool {
...
+		return value
+	}
+}
+
+// ConstLabels returns a value provider for a constant.
+func ConstLabels(labels map[string]string) func() map[string]string {
+	return func() map[string]string {
+		return labels
+	}
+}
```
Comment (dhermes) [inline cron/util.go]: These 4 don't seem like they belong in `cron` pkg but I suppose a little copying is better than a little dependency

File: r2/opt_log.go [hunk 15]
```diff
@@ -1,23 +1,18 @@
...
+	return func(r *Request) error {
+		if err := OptLogRequest(log)(r); err != nil {
+			return err
+		}
+		if err := OptLogResponse(log)(r); err != nil {
+			return err
+		}
 		return nil
-	})
+	}
```
Comment (dhermes) [inline r2/opt_log.go]: This seems **very** unrelated to the PR topic.

File: web/static_file_server.go (hunk 7 above)
Comment (dhermes) [reply web/static_file_server.go]: This is also (possibly) relevant to logging with `dd.trace_id` and `dd.span_id`

File: webutil/event_source.go [hunk 16]
```diff
@@ -3,72 +3,98 @@ package webutil
...
+func (es *EventSource) Ping() error {
 	return es.Event("ping")
 }
 
 // Event writes an event.
-func (es EventSource) Event(name string) error {
-	_, err := io.WriteString(es.Output, "event: "+name+"\n\n")
+func (es *EventSource) Event(name string) error {
+	es.Lock()
+	defer es.Unlock()
```
Comment (dhermes) [inline webutil/event_source.go]: Locking to protect `es.Output` from races / simultaneous writes?

File: webutil/event_source.go [hunk 17]
```diff
@@ -3,72 +3,98 @@ package webutil
...
 	}
-	if typed, ok := es.Output.(http.Flusher); ok {
+	return es.dataUnsafe(data)
+}
+
+//
+// unsafe methods
+//
+
+func (es *EventSource) pingUnsafe() error {
```
Comment (dhermes) [inline webutil/event_source.go]: This helper seems unneeded?

File: async/latch.go (hunk 2 above)
Comment (wcharczuk) [reply async/latch.go]: yeah i'll do that now.

File: async/autoflush_buffer.go [hunk 18]
```diff
@@ -65,7 +65,7 @@ func OptAutoflushBufferFlushOnStop(flushOnStop bool) AutoflushBufferOption {
 // A handler should be provided but without one the buffer will just clear.
 // Adds that would cause fixed length flushes do not block on the flush handler.
 type AutoflushBuffer struct {
-	*Latch
+	Latch       *Latch
```
[the remark this answers is not in the document]
Comment (wcharczuk) [reply async/autoflush_buffer.go]: embedding adds a bunch of stuff that can get shadow overwritten and it's harder to reason about

File: async/batch.go [hunk 19]
```diff
@@ -46,27 +46,34 @@ type Batch struct {
...
 	for x := 0; x < b.Parallelism; x++ {
 		worker := NewWorker(b.Action)
 		worker.Errors = b.Errors
 		worker.Finalizer = returnWorker
+
+		workerStarted := worker.NotifyStarted()
 		go worker.Start()
-		<-worker.NotifyStarted()
-		workers <- worker
+		<-workerStarted
```
[the remark this answers is not in the document]
Comment (wcharczuk) [reply async/batch.go]: yeah kind of

File: async/batch.go [hunk 20]
```diff
@@ -46,27 +46,34 @@ type Batch struct {
...
+		availableWorkers <- worker
 	}
 
 	defer func() {
-		for x := 0; x < b.Parallelism; x++ {
-			worker := <-workers
-			worker.Stop()
+		// stop the workers
+		for x := 0; x < len(allWorkers); x++ {
+			allWorkers[x].Stop()
```
[the remark this answers is not in the document]
Comment (wcharczuk) [reply async/batch.go]: it did in rare occasions / unit tests

File: async/batch_test.go [hunk 21]
```diff
@@ -25,8 +25,37 @@ func TestBatch(t *testing.T) {
...
 
 	assert.Equal(32, processed)
 	assert.Equal(32, len(errors))
 }
+
+func TestBatchPanic(t *testing.T) {
+	assert := assert.New(t)
+
+	items := make(chan interface{}, 32)
+	for x := 0; x < 32; x++ {
```
[the remark this answers is not in the document]
Comment (wcharczuk) [reply async/batch_test.go]: i'm going to use `l` and `I` just for you from now on

File: async/batch_test.go [hunk 22]
```diff
@@ -25,8 +25,37 @@ func TestBatch(t *testing.T) {
...
+		items <- "hello" + strconv.Itoa(x)
+	}
+
+	var processed int32
+	action := func(_ context.Context, v interface{}) error {
+		atomic.AddInt32(&processed, 1)
+		if processed == 1 {
```
[the remark this answers is not in the document]
Comment (wcharczuk) [reply async/batch_test.go]: yeah that's fair

File: r2/opt_log.go (hunk 15 above)
Comment (wcharczuk) [reply r2/opt_log.go]: yeah this became a bit of a grab bag

File: webutil/event_source.go (hunk 16 above)
Comment (wcharczuk) [reply webutil/event_source.go]: yeah exactly, it makes it synchronized by default which isn't necessarily always a good move but for this felt like it would be wise

File: webutil/event_source.go (hunk 17 above)
Comment (wcharczuk) [reply webutil/event_source.go]: yeah it's fine just to copy "ping"

File: async/batch_test.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (dhermes) [reply async/batch_test.go]: Haha, no need

File: async/latch.go [hunk 23]
```diff
@@ -37,6 +37,17 @@ type Latch struct {
...
+	l.Lock()
+	l.state = LatchStopped
+	l.starting = make(chan struct{}, 1)
+	l.started = make(chan struct{}, 1)
+	l.stopping = make(chan struct{}, 1)
+	l.stopped = make(chan struct{}, 1)
```
[the remark this answers is not in the document]
Comment (wcharczuk) [reply async/latch.go]: The intent is that, in a normal lifecycle, you can have each stage be hit (1) time, and that the channels won't necessarily block, so it follows this would be hard  coded.

File: async/latch.go [hunk 24]
```diff
@@ -106,56 +117,40 @@ func (l *Latch) NotifyStopped() (notifyStopped <-chan struct{}) {
...
-	l.Lock()
-	defer l.Unlock()
-
 	if l.IsStarting() {
 		return
 	}
 	atomic.StoreInt32(&l.state, LatchStarting)
-	close(l.starting)
-	l.starting = make(chan struct{})
+	l.starting <- struct{}{}
```
[the remark this answers is not in the document]
Comment (wcharczuk) [reply async/latch.go]: yeah that's the tension with the existing use case; the multiple listeners aspect imho wasn't worth the really nasty race conditions it caused.

File: async/latch_test.go [hunk 25]
```diff
@@ -63,19 +63,4 @@ func TestLatch(t *testing.T) {
...
-
-	assert.True(l.IsStarting())
-	assert.False(l.IsStarted())
-	assert.False(l.IsStopping())
-	assert.False(l.IsStopped())
-
-	go func() {
-		l.Started()
-	}()
-	<-l.NotifyStarted()
```
[the remark this answers is not in the document]
Comment (wcharczuk) [reply async/latch_test.go]: that section of the test isn't relevant w/ the `.Reset()` change

File: async/worker.go [hunk 26]
```diff
@@ -123,10 +121,14 @@ func (qw *Worker) Drain(ctx context.Context) {
 	<-stopped
 }
 
-// Close stops the worker and cleans up resources.
-func (qw *Worker) Close() error {
```
[the remark this answers is not in the document]
Comment (wcharczuk) [reply async/worker.go]: No reason to have it w/ `Stop()` and `Drain()`

File: async/worker.go [hunk 27]
```diff
@@ -123,10 +121,14 @@ func (qw *Worker) Drain(ctx context.Context) {
...
-	close(qw.Work)
-	return nil
+// HandleError sends a non-nil err to the error
+// collector if one is provided.
+func (qw *Worker) HandleError(err error) {
+	if err == nil {
+		return
+	}
+	if qw.Errors == nil {
+		return
```
[the remark this answers is not in the document]
Comment (wcharczuk) [reply async/worker.go]: 1) you don't want background processes to panic ever
2) qw.Errors is set in the ctor and you never (or shouldn't) have races around the reference after that
3) that's what the errors channel is for

File: cron/job_manager.go (hunk 10 above)
Comment (wcharczuk) [reply cron/job_manager.go]: yeah exactly

File: cron/job_manager.go (hunk 11 above)
Comment (wcharczuk) [reply cron/job_manager.go]: they're done in `.Load()` and are duplicative

File: cron/job_manager.go (hunk 12 above)
Comment (wcharczuk) [reply cron/job_manager.go]: Yeah they're a huge overlap with `.Start()` and `.Stop()` and originally were intended to be there to preserve goroutines that had already started but in practice goroutines are so cheap it's better just to not have this code.

File: cron/job_manager_test.go (hunk 13 above)
Comment (wcharczuk) [reply cron/job_manager_test.go]: yeah someday the great `s/assert := assert.New(t)/it := assert.New(t)` will happen but today is not that day

File: cron/util.go (hunk 14 above)
Comment (wcharczuk) [reply cron/util.go]: it's more because of the nature of these specific predicates which is kind of specific to this use case

File: async/worker.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (dhermes) [reply async/worker.go]: 1. (I wasn't suggesting it was a **good** option, just an option)

File: async/latch.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (dhermes) [reply async/latch.go]: So let's update the godoc on all of the `Notify*()` helpers to indicate they are one time use? You **could** go one step further and mark them as "taken" so if a `Notify*()` gets invoked again it'd error.

File: async/latch.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (dhermes) [reply async/latch.go]: This was more like a software hygiene thing, so that the "fresh" state that `Reset()` thinks it needs is the same as the "fresh" state that `NewLatch()`

**Comment by dhermes** [comment]
OK AFAICT just 3 dangling things to wrap up before merging:

- Updating godoc that notifies are single use ([here][1])
- Dangling question about why locking has been removed on `Latch` helpers ([here][2])
- (Optional / I can give on this): Share code between `Latch.Reset()` and `NewLatch()` ([here][3])

[1]: https://github.com/blend/go-sdk/pull/160/files#r365457422
[2]: https://github.com/blend/go-sdk/pull/160/files#r364932515
[3]: https://github.com/blend/go-sdk/pull/160/files#r365459988

File: async/latch.go [hunk 28]
```diff
@@ -106,56 +117,40 @@ func (l *Latch) NotifyStopped() (notifyStopped <-chan struct{}) {
 // Starting signals the latch is starting.
 // This is typically done before you kick off a goroutine.
 func (l *Latch) Starting() {
-	l.Lock()
-	defer l.Unlock()
```
[the remark this answers is not in the document]
Comment (wcharczuk) [reply async/latch.go]: it was needed before because we were mutating a field on the latch, now we're just pushing data into an already synchronized structure (a channel) and using an atomic set to update the state

**Comment by dhermes** [comment]
FYI @wcharczuk I rebased this after #161 merged (trying to clean up my own disruption). If your local branch won't push because of "fast forward" issues, that's why.

File: cron/job_scheduler.go [hunk 29]
```diff
@@ -109,60 +33,138 @@ func NewJobScheduler(job Job, options ...JobSchedulerOption) *JobScheduler {
...
 type JobScheduler struct {
 	sync.Mutex
-	Latch *async.Latch
 
 	Job    Job
 	Config JobConfig
 	Tracer Tracer
 	Log    logger.Log
 
-	Schedule    Schedule
```
Comment (dhermes) [inline cron/job_scheduler.go]: For future readers of this PR, this changed caused a regression for **stateful** `Schedule` instances. For such instances, users (i.e. **me**) were relying on `JobScheduler` to manage the state of the their schedule (one could argue if this is something that **should** be done, but it was a feature being utilized, hello [Hyrum's Law][1] 👋).

The regression was addressed in #172.

[1]: https://www.hyrumslaw.com/
